---
title: "RNAseq Differentially Expressed Genes analysis using DESeq2 and edgeR"
author: "Stephanie Ali Fairbairn"
date: "30/04/2021"
output: html_document
---
**Biology-722 Project**
  
```{r, eval= FALSE}
#Install all required packages
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager", repos = "https://cran.rstudio.com")

BiocManager::install("Rbowtie2")
BiocManager::install("Rsamtools")
install.packages("Rtools", repos = "https://cran.rstudio.com")
install.packages("GREP2", repos = "https://cran.rstudio.com")
install.packages("edgeR", repos = "https://cran.rstudio.com")
install.packages("UpSetR", repos = "https://cran.rstudio.com")
```

```{r, eval= FALSE}
#Set working directory
setwd("E:/Personal-Project/RCode") # <-- edit this based on the directory of your choice
getwd()
```

```{r, eval= FALSE}
#Load all required libraries
library(Rbowtie2)
library(Rsamtools)
library(R.utils)
library(Ape)
library(Rsubread)
library(DESeq2)
library(edgeR)
library(ggplot2)
library(ggrepel)
library(plyr)
library(VennDiagram)
library(Biobase)
library(dplyr)
library(UpSetR)
```

```{r, eval= FALSE}
# Creating objects for other working directories   <-- edit based on your directories of choice
td1 <- "E:/RNAseq-Raw-Data/Reference-genome"
td2 <- "E:/RNAseq-Raw-Data"
td3 <- "E:/RNAseq-Raw-Data/Sam_Out"
td4 <- "E:/RNAseq-Raw-Data/Bam_Files"
td5 <- "E:/RNAseq-Raw-Data/Final_Results"
td_64A2 <- "E:/RNAseq-Raw-Data/64A2_APC"
td_64A3 <- "E:/RNAseq-Raw-Data/64A3_AK"
td_66A1 <- "E:/RNAseq-Raw-Data/66A1_APC"
td_66A3 <- "E:/RNAseq-Raw-Data/66A3_APC"
td_67A1 <- "E:/RNAseq-Raw-Data/67A1_APC"
td_68A5 <- "E:/RNAseq-Raw-Data/68A5_AK"
td_69A1 <- "E:/RNAseq-Raw-Data/69A1_AK"
td_71A1 <- "E:/RNAseq-Raw-Data/71A1_AK"
td_list <- list(td_64A2, td_64A3, td_66A1, td_66A3, td_67A1, td_68A5, td_69A1, td_71A1)
```

```{r, eval= FALSE}
# All .gz Fastq data files were unzipped using the following
setwd(td_64A2)
gunzip("64A2_APC_R1.fastq.gz")
gunzip("64A2_APC_R2.fastq.gz")
setwd(td_64A3)
gunzip("64A3_AK_R1.fastq.gz")
gunzip("64A3_AK_R2.fastq.gz")
setwd(td_66A1)
gunzip("66A1_APC_R1.fastq.gz")
gunzip("66A1_APC_R2.fastq.gz")
setwd(td_66A3)
gunzip("66A3_APC_R1.fastq.gz")
gunzip("66A3_APC_R2.fastq.gz")
setwd(td_67A1)
gunzip("67A1_APC_R1.fastq.gz")
gunzip("67A1_APC_R2.fastq.gz")
setwd(td_68A5)
gunzip("68A5_AK_R1.fastq.gz")
gunzip("68A5_AK_R2.fastq.gz")
setwd(td_69A1)
gunzip("69A1_AK_R1.fastq.gz")
gunzip("69A1_AK_R2.fastq.gz")
setwd(td_71A1)
gunzip("71A1_AK_R1.fastq.gz")
gunzip("71A1_AK_R2.fastq.gz")
```

```{r, eval= FALSE}
## Copy all .fastq files from local to remote terminal in Mobaxterm to perform fastqc
## scp -r <filename>.fastq gradstd1@info.mcmaster.ca:~
## e.g. <filename> = 64A2_APC_R1

# Perform fastqc on all .fastq files and obtain adapter sequences in Mobaxterm
## fastqc *.fastq -o .

# Copy all fastqc files from remote to local terminal in Mobaxterm
## scp -r gradstd1@info.mcmaster.ca:~/*fastqc* .

##The fastqc zipped files were provided for your use. 

# Running cutadapt with **Python** using MSDOS Command Prompt in Windows
## Using Illumina True seq genome sequence adapter "AGATCGGAAGAGCACACGTCTGAACTCCAGTCA" for forward reads, and
## genome sequence adapter "AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT" for reverse reads
## py -m cutadapt -a AGATCGGAAGAGCACACGTCTGAACTCCAGTCA -A AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT -o <filename>_R1_trimmed.fastq -p <filename>_R2_trimmed.fastq <filename>_R1.fastq <filename>_R2.fastq --cores=0 --minimum-length 20
## e.g. <filename> = 64A2_APC

# Mouse annotation and genome reference files were downloaded from (gencodegenes.org)
## gencode.vM26.primary_assembly.annotation.gtf.gz
## GRCm39.primary_assembly.genome.fa.gz
```

```{r, eval= FALSE}
# Files were unzipped using gunzip function
setwd(td1)
gunzip("gencode.vM26.primary_assembly.annotation.gtf.gz")
gunzip("GRCm39.primary_assembly.genome.fa.gz")
```

```{r, eval= FALSE}
** Aligning sequence reads to a reference genome**
  
  # Building reference index using Bowtie2
  bowtie2_build(references="GRCm39.primary_assembly.genome.fa", bt2Index = "musculus_bowtie")

# Aligning sequences to the reference genome for all sequence reads

bowtie2(bt2Index = file.path(td1, "musculus_bowtie"), 
        samOutput = file.path(td3, "sam-out-64A2_APC.sam"), 
        seq1 = file.path(td_64A2, "64A2_APC_R1_trimmed.fastq"), 
        seq2 = file.path(td_64A2, "64A2_APC_R2_trimmed.fastq"),
        overwrite = TRUE,
        "--threads 8")

bowtie2(bt2Index = file.path(td1, "musculus_bowtie"), 
        samOutput = file.path(td3, "sam-out-64A3_AK.sam"), 
        seq1 = file.path(td_64A3, "64A3_AK_R1_trimmed.fastq"), 
        seq2 = file.path(td_64A3, "64A3_AK_R2_trimmed.fastq"),
        overwrite = TRUE,
        "--threads 8")

bowtie2(bt2Index = file.path(td1, "musculus_bowtie"), 
        samOutput = file.path(td3, "sam-out-66A1_APC.sam"), 
        seq1 = file.path(td_66A1, "66A1_APC_R1_trimmed.fastq"), 
        seq2 = file.path(td_66A1, "66A1_APC_R2_trimmed.fastq"),
        overwrite = TRUE,
        "--threads 8")

bowtie2(bt2Index = file.path(td1, "musculus_bowtie"), 
        samOutput = file.path(td3, "sam-out-66A3_APC.sam"), 
        seq1 = file.path(td_66A3, "66A3_APC_R1_trimmed.fastq"), 
        seq2 = file.path(td_66A3, "66A3_APC_R2_trimmed.fastq"),
        overwrite = TRUE,
        "--threads 8")

bowtie2(bt2Index = file.path(td1, "musculus_bowtie"), 
        samOutput = file.path(td3, "sam-out-67A1_APC.sam"), 
        seq1 = file.path(td_67A1, "67A1_APC_R1_trimmed.fastq"), 
        seq2 = file.path(td_67A1, "67A1_APC_R2_trimmed.fastq"),
        overwrite = TRUE,
        "--threads 8")

bowtie2(bt2Index = file.path(td1, "musculus_bowtie"), 
        samOutput = file.path(td3, "sam-out-68A5_AK.sam"), 
        seq1 = file.path(td_68A5, "68A5_AK_R1_trimmed.fastq"), 
        seq2 = file.path(td_68A5, "68A5_AK_R2_trimmed.fastq"),
        overwrite = TRUE,
        "--threads 8")

bowtie2(bt2Index = file.path(td1, "musculus_bowtie"), 
        samOutput = file.path(td3, "sam-out-69A1_AK.sam"), 
        seq1 = file.path(td_69A1, "69A1_AK_R1_trimmed.fastq"), 
        seq2 = file.path(td_69A1, "69A1_AK_R2_trimmed.fastq"),
        overwrite = TRUE,
        "--threads 8")

bowtie2(bt2Index = file.path(td1, "musculus_bowtie"), 
        samOutput = file.path(td3, "sam-out-71A1_AK.sam"), 
        seq1 = file.path(td_71A1, "71A1_AK_R1_trimmed.fastq"), 
        seq2 = file.path(td_71A1, "71A1_AK_R2_trimmed.fastq"),
        overwrite = TRUE,
        "--threads 8")
```

```{r, eval= FALSE}
**Using featureCounts to generate a count table**
  
  # Create list with all .sam files for input into featureCounts
  setwd(td3)
sam_files <- dir(".", "sam$")

# Create variable for mouse annotation filename
setwd(td1)
annotation_gtf <- "gencode.vM26.primary_assembly.annotation.gtf"

# Perform featureCounts on all .sam files using mouse annotation file
setwd(td3)
sorted_counts <- featureCounts(files=sam_files, annot.ext=file.path(td1, annotation_gtf), isGTFAnnotationFile=TRUE, GTF.featureType="gene", GTF.attrType="gene_id", useMetaFeatures = FALSE, isPairedEnd = TRUE, nthreads = 8)
head(sorted_counts)

# Create counts data file from featureCounts output for use in DESeq2 Analysis and EdgeR Analysis 
counts <- sorted_counts$counts
head(counts)
```

```{r, eval= FALSE}
** Perform DESeq2 Analysis to obtain differentially expressed genes**
  
  # Check order of sam files in counts data
  
  # Create genotype vector based on order of sam files to be used in DESeq2 analysis
  genotype <- c("APC", "AK", "APC", "APC", "APC", "AK", "AK", "AK")

# Create matrix aligning genotype with .sam files
coldata <- data.frame(sam_files, genotype)

# Preparing to perform DESeq2 Analysis using featureCounts output
# Create DESeq2 dataset prior to DESeq analysis
deseqdata <- DESeqDataSetFromMatrix(countData = counts, colData = coldata, design = ~genotype)

# Adding genotype vector to DESeq2 dataset
deseqdata$genotype <- factor(deseqdata$genotype, levels = c("APC","AK"))

# Perform DESeq2 analysis on dataset
deseqdata <- DESeq(deseqdata)

# Extract the deseq_results table from the DESeq2 analysis
deseq_results <- results(deseqdata)
deseq_results

# Reorder the deseq_results based on the "padj" values
deseq_ordered <- deseq_results[order(deseq_results$padj),]
deseq_ordered

# Remove all rows with padj = NA
deseq_final <- deseq_ordered[!is.na(deseq_ordered$padj), ]
deseq_final
summary(deseq_final)
dim(deseq_final)

# Assigning padj and log2FoldChange limits to obtain DEGs 
pval_threshold <- 0.05
logfc_threshold <- 1.5

# Convert S4 object to dataframe
deseq_final_df <- as.data.frame(deseq_final)

# Add column for gene id
deseq_final_df$genes <- row.names(deseq_final_df)
```

```{r, eval= FALSE}
** Creating a Volcano plot to show DEGs**
  # add a column of "Not Sig"
  deseq_final_df$Legend <- "Not Sig"
# if log2Foldchange > 1.5 and padj < 0.05, set as "Upreg" 
deseq_final_df$Legend[deseq_final_df$log2FoldChange > log2(logfc_threshold) & deseq_final_df$padj < pval_threshold] <- "Upreg"
# if log2Foldchange < -1.5 and padj < 0.05, set as "Downreg"
deseq_final_df$Legend[deseq_final_df$log2FoldChange < -log2(logfc_threshold) & deseq_final_df$padj < pval_threshold] <- "Downreg"
summary(deseq_final_df)

#Creating subsets of the final results dataframe to be used later for plots
deseq_final_subset1 <- subset(deseq_final_df, Legend == "Upreg")
deseq_final_subset2 <- subset(deseq_final_df, Legend == "Downreg")
deseq_final_subset3 <- subset(deseq_final_df, Legend == "Not Sig")

#Create a subset of both Upreg and Downreg genes to be used later for plots
deseq_final_subset4 <- subset(deseq_final_df, Legend != "Not Sig")

# Get the rownames for significantly expressed genes for use later on in plots
deseq_row_names <- row.names(deseq_final_subset4)

# Create a named vector: the values are the colors to be used, 
# the names are the categories they will be assigned to
mycolors <- c("red", "green", "grey")
names(mycolors) <- c("Downreg", "Upreg", "Not Sig")

# Creating plot of results for AK vs APC for all DEGs according to DESeq2
g_deseq = ggplot(deseq_final_df, aes(x=log2FoldChange, y=-log10(padj)), colour=Legend) +
  geom_point(aes(x=log2FoldChange, y=-log10(padj), colour=Legend)) +
  geom_text_repel(aes(label=ifelse(padj < 0.1e-30 & abs(log2FoldChange) >= logfc_threshold, row.names(deseq_final_df), ''))) +
  ggtitle("AK vs APC DEGs according to DESeq2") +
  xlab("log2 fold change") + 
  ylab("-log10 adjusted p-value") +
  theme_bw() +
  scale_colour_manual(values = mycolors) +
  theme(legend.position = "right", plot.title = element_text(size = rel(1.5), hjust = 0.5), axis.title = element_text(size = rel(1.25)))


# Plot graph
g_deseq

# End of DESeq2 Analysis
```

```{r, eval= FALSE}
**Preparing to perform edgeR Analysis using featureCounts output**
  
  # Create genotype vector group based on order of sam files
  Group = rep(c("APC", "AK", "APC", "APC", "APC", "AK", "AK", "AK")) 

# Setting up the model
Group <- factor(c(1,2,1,1,1,2,2,2))

# Creating a DGEList object to hold the read counts
edgeR_dgList <- DGEList(counts = counts, genes = rownames(counts), group = Group)
edgeR_dgList
edgeR_dgList$samples
head(edgeR_dgList$genes)

# Filtering sample to retain only genes that are represented at
# least 1 count per million reads in at least two samples
countsPerMillion <- cpm(edgeR_dgList)
summary(countsPerMillion)
countCheck <- countsPerMillion > 1
keep <- which(rowSums(countCheck) >= 2)
edgeR_dgList <- edgeR_dgList[keep,]
summary(cpm(edgeR_dgList))

# Normalising RNA-seq both within and between samples
# edgeR implements the trimmed mean of M-values (TMM) method
edgeR_dgList <- calcNormFactors(edgeR_dgList, method = "TMM")
summary(cpm(edgeR_dgList))


#  Specifying the design matrix
designMat <- model.matrix(~Group)
designMat

# Estimating the dispersion parameters for the negative binomial model
# Using empirical Bayes method to shrink the genewise dispersion estimates
# towards the common dispersion (tagwise dispersion)
edgeR_dgList <- estimateGLMCommonDisp(edgeR_dgList, design=designMat)
edgeR_dgList <- estimateGLMTrendedDisp(edgeR_dgList, design=designMat)
edgeR_dgList <- estimateGLMTagwiseDisp(edgeR_dgList, design=designMat)
edgeR_dgList$common.dispersion
edgeR_dgList$trended.dispersion
edgeR_dgList$tagwise.dispersion

# Finding differentially expressed genes
edgeR_fit <- glmFit(edgeR_dgList, designMat)
head(edgeR_fit)

summary(edgeR_fit$coefficients)
edgeR_lrt <- glmLRT(edgeR_fit, coef = 2)

# Create dataframe of edgeR_lrt from the glmLRT output
edgeR_results <- as.data.frame(topTags(edgeR_lrt,n=nrow(edgeR_lrt)))
summary(edgeR_results)
dim(edgeR_results)
```

```{r, eval= FALSE}
** Creating a volcano plot to show DGEs**
  
  # Setting FDR and logFC limits to obtain DEGs
  FDR_threshold <- 0.05
logfc_threshold <- 1.5

# add a column of "Not Sig"
edgeR_results$Legend <- "Not Sig"
# if logFC > 1.5 and FDR < 0.05, set as "Upreg" 
edgeR_results$Legend[edgeR_results$logFC > log2(logfc_threshold) & edgeR_results$FDR < FDR_threshold] <- "Upreg"
# if logFC < -1.5 and FDR < 0.05, set as "Downreg"
edgeR_results$Legend[edgeR_results$logFC < -log2(logfc_threshold) & edgeR_results$FDR < FDR_threshold] <- "Downreg"

#Creating subsets of the results dataframe to be used later on for plots
edgeR_results_subset1 <- subset(edgeR_results, Legend == "Upreg")
edgeR_results_subset2 <- subset(edgeR_results, Legend == "Downreg")
edgeR_results_subset3 <- subset(edgeR_results, Legend == "Not Sig")
#Create a subset of both Upreg and Downreg genes to be used later for plots
edgeR_results_subset4 <- subset(edgeR_results, Legend != "Not Sig")

# Get the rownames for significantly expressed for use later on in plots
edgeR_row_names <- row.names(edgeR_results_subset4)

# Create a named vector: the values are the colors to be used, 
# the names are the categories they will be assigned to
mycolors <- c("red", "green", "grey")
names(mycolors) <- c("Downreg", "Upreg", "Not Sig")

# Creating plot of results for AK vs APC for all DEGs according to edgeR
g_edgeR = ggplot(edgeR_results, aes(x=logFC, y=-log10(FDR)), colour=Legend) +
  geom_point(aes(x=logFC, y=-log10(FDR), colour=Legend)) +
  geom_text_repel(aes(label=ifelse(FDR < 0.1e-30 & abs(logFC) >= logfc_threshold, row.names(edgeR_results), ''))) +
  ggtitle("AK vs APC DEGs according to edgeR") +
  xlab("log2 fold change") + 
  ylab("-log10 adjusted FDR") +
  theme_bw() +
  scale_colour_manual(values = mycolors) +
  theme(legend.position = "right", plot.title = element_text(size = rel(1.5), hjust = 0.5), axis.title = element_text(size = rel(1.25)))


# Plot graph
g_edgeR

# End of edgeR Analysis
```

```{r, eval= FALSE}
**Creating a Venn diagram to show overlap**
  
  # Calculating the intersection of the two sets of results from DESeq2 and edgeR to show overlap
  deg_intersect = length(intersect(edgeR_row_names, deseq_row_names))
deg_venn <- list('intersect' = deg_intersect,
                 'edgeR' = length(edgeR_row_names),
                 'DESeq2' = length(deseq_row_names))

# Arguments for a pairwise (two-sets) venn-diagram are sizes for edgeR, DESeq2 and overlap (intersect)
# Plotting venn diagram
venn_plot <- draw.pairwise.venn(deg_venn$edgeR, deg_venn$DESeq2, deg_venn$intersect,
                                category = c("edgeR", "DESeq2"), cat.cex = 1.5, cex= 1.5, scaled = F,
                                fill = c("medium orchid", "sky blue"), alpha = rep(0.5, 2),
                                cat.pos = c(0, 0))

# Adding label to venn diagram
require(gridExtra)
grid.arrange(gTree(children = venn_plot), top = textGrob("Venn diagram of SE genes", gp=gpar(fontsize=25)))
```

```{r, eval= FALSE}
# Preparing upset plot
DE_list <- list(edgeR = rownames(subset(edgeR_results_subset4, FDR <= FDR_threshold)), DESeq2 = rownames(subset(deseq_final_subset4, padj <= pval_threshold)))
DE_gns <- fromList(DE_list)
summary(DE_gns)
upset(DE_gns, order.by = "freq")
```

```{r, eval= FALSE}
# Creating pairwise plot
DE_df <- merge(edgeR_results_subset4, deseq_final_subset4, by="genes", all=TRUE)
DE_fc <- data.frame(edgeR = DE_df$logFC, DESeq2 = DE_df$log2FoldChange)
pairs(DE_fc)
```

```{r, eval= FALSE}
# Comparison of differentially expressed genes from results of DESeq2 and edgeR analysis
merged <- merge(deseq_final_df, edgeR_results, by='genes', all= TRUE)
with(merged, plot(logFC, log2FoldChange, xlab="logFC edgeR", ylab="log2FC DESeq2", pch=20, col="black", main="Fold change for DESeq2 vs edgeR"))
with(subset(merged, FDR<FDR_threshold),  points(logFC, log2FoldChange, pch=20, col="red"))
with(subset(merged, padj<pval_threshold), points(logFC, log2FoldChange, pch=20, col="green"))
legend("topleft", xjust=1, yjust=1, legend=c("FDR<0.05 edgeR only", "FDR<0.05 DESeq & edgeR", "FDR>0.05"), pch=20, col=c("red", "green", "black"), bty="n", y.intersp = .25)
```

```{r, eval= FALSE}
# Getting summaries of logFCs for DESeq2 and edgeR
summary(deseq_final_subset4$log2FoldChange)
summary(edgeR_results_subset4$logFC)

# Getting summaries of the adjusted pvalues for DESeq2 and edgeR
summary(deseq_final_subset4$padj)
summary(edgeR_results_subset4$FDR)
```

```{r, eval= FALSE}
# Comparison of fold change for significantly expressed genes
boxplot(DE_df$log2FoldChange, DE_df$logFC, main= "Foldchange comparison", names= c("DESeq2","edgeR"), horizontal=TRUE, col= c("sky blue","medium orchid"), las= 2, style="quantile", notch=TRUE)

# Comparison of adjusted Pvalue for significantly expressed genes
boxplot(DE_df$padj, DE_df$FDR, main= "Adjusted Pvalue comparison", names= c("DESeq2","edgeR"), horizontal=TRUE, col= c("sky blue","medium orchid"), las= 2, style="tukey", notch=TRUE)
```

```{r, eval= FALSE}
#Creating a histogram comparing fold change for both methods
edgeR_FC <- DE_df$logFC
deseq_FC <- DE_df$log2FoldChange
hist(edgeR_FC, breaks=200, col = "red",xlab = "Fold change", ylab = "Frequency", main = "Fold change Comparison")
hist(deseq_FC, breaks=200, col = "sky blue", add=T)
legend("topright", c("edgeR","DESeq2"), col = c("red", "sky blue"), lwd = 2, bty="n", y.intersp = .25)

#Creating a histogram comparing adjusted Pvalue for both methods
edgeR_FC <- DE_df$padj
deseq_FC <- DE_df$FDR
hist(edgeR_FC, breaks=200, col = "red", xlab = "Adjusted p value", ylab = "Frequency", main = "Adjusted Pvalue Comparison")
hist(deseq_FC, breaks=200, col = "sky blue", add=T)
legend("topright", c("edgeR","DESeq2"), col = c("red", "sky blue"), lwd = 2, bty="n", y.intersp = .25)
```

```{r, eval= FALSE}
** Creating tables to show top 20 upregulated and downregulated genes**
  
  # Create temporary dataframe for DESeq2 SE genes
  temp1 <- deseq_final_subset4
names(temp1)[names(temp1) == "Legend"] <- "Expression"
setwd(td5)

# Top 20 Upreg genes for DESeq2 based on highest log2FoldChange values
temp2 <- subset(temp1, Expression == "Upreg")
temp2 <- temp2[order(temp2$log2FoldChange, decreasing=TRUE),]
temp2 <- temp2[1:20, ]
Top20_Upreg_log2FC_deseq <- select(temp2, "genes", "padj", "log2FoldChange", "Expression")
write.csv(Top20_Upreg_log2FC_deseq, "Top20_Upreg_log2FC_deseq.csv")

# Top 20 Downreg genes for DESeq2 based on lowest log2FoldChange values
temp2 <- subset(temp1, Expression == "Downreg")
temp2 <- temp2[order(temp2$log2FoldChange),]
temp2 <- temp2[1:20, ]
Top20_Downreg_log2FC_deseq <- select(temp2, "genes", "padj", "log2FoldChange", "Expression")
write.csv(Top20_Downreg_log2FC_deseq, "Top20_Downreg_log2FC_deseq.csv")


# Create temporary dataframe for edgeR SE genes
temp1 <- edgeR_results_subset4
names(temp1)[names(temp1) == "Legend"] <- "Expression"
setwd(td5)

# Top 20 Upreg genes for edgeR based on highest logFC values
temp2 <- subset(temp1, Expression == "Upreg")
temp2 <- temp2[order(temp2$logFC, decreasing=TRUE),]
temp2 <- temp2[1:20, ]
Top20_Upreg_logFC_edgeR <- select(temp2, "genes", "FDR", "logFC", "Expression")
write.csv(Top20_Upreg_logFC_edgeR, "Top20_Upreg_logFC_edgeR.csv")

# Top 20 Downreg genes for edgeR based on lowest logFC values
temp2 <- subset(temp1, Expression == "Downreg")
temp2 <- temp2[order(temp2$logFC),]
temp2 <- temp2[1:20, ]
Top20_Downreg_logFC_edgeR <- select(temp2, "genes", "FDR", "logFC", "Expression")
write.csv(Top20_Downreg_logFC_edgeR, "Top20_Downreg_logFC_edgeR.csv")
```

